### PR TITLE
Add HTTPServer so we don't vend BlueSocketHTTP

### PR DESCRIPTION
**Pull request by ianpartridge**
This PR adds a simple `HTTPServer` protocol, and an implementation of it called `SimpleHTTPServer`. The purpose of this is so that people can use this package to run an HTTP server without having to refer to BlueSocket related types. In the future, we want to remove all of those from the package and we can't do that if we've vended them as public API.

It also renames the `WebApp` and `WebAppContaining` types (which didn't have much support) to `HTTPRequestHandler` and `HTTPRequestHandling`.

### CODE REVIEW

**Comment by seabaylea** [comment]
@swift-server/stakeholders any opinions of this as a very early implementation of an API to create/start the HTTP Server itself?

File: Sources/HTTP/HTTPServer.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+//
+
+/// Definition of an HTTP server.
+public protocol HTTPServer : class {
+
+    /// Start the HTTP server on the given `port`, using `handler` to process incoming requests
+    func start(port: Int, handler: @escaping HTTPRequestHandler) throws
```
Comment (tanner0101) [inline Sources/HTTP/HTTPServer.swift]: `bind` also accepts an address name. should we support that?

**Review by tanner0101 — APPROVED** [review]
LGTM

File: Sources/HTTP/HTTPServer.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    var port: Int { get }
+
+    /// The number of current connections
+    var connectionCount: Int { get }
+}
+
+/// A basic HTTP server. Currently this is implemented using the BlueSocket
+/// abstraction, but the intention is to remove this dependency and reimplement
+/// the class using transport APIs provided by the Server APIs working group.
+public class SimpleHTTPServer {
```
Comment (adellibovi) [inline Sources/HTTP/HTTPServer.swift]: I am not sure to understand the purpose of this public class. If it is something that devs are supposed to use, wouldn't it be better to call the protocol `HTTPServerType` and this class just `HTTPServer`?

**Review by adellibovi — APPROVED** [review]
Nice 👏 !

File: Sources/HTTP/HTTPServer.swift (hunk 2 above)
Comment (seabaylea) [reply Sources/HTTP/HTTPServer.swift]: This needs to have a public constructor. And agreed - I'd prefer the concrete server class to be just `HTTPServer`

File: Sources/HTTP/HTTPServer.swift (hunk 1 above)
Comment (ianpartridge) [reply Sources/HTTP/HTTPServer.swift]: Golang's [net/http/Server](https://golang.org/pkg/net/http/#Server) has an `Addr string` field which seems to encapsulate both the address name and port. I'm not sure whether there's a standard format for this that we could/should use?

**Comment by ianpartridge** [comment]
I've updated the PR to rename `SimpleHTTPServer` to `HTTPServer`, and I've added a default initializer to `HTTPServer`. More feedback very welcome.

**Comment by ianpartridge** [comment]
Question for people: do we want to mark `HTTPServer` as `final`?

**Comment by aleksey-mashanov** [comment]
Personally, I not really sure what universal API for initializing http servers should exist. Of course it can cover simple handmade http servers (maybe even both sync and async), but advanced configuration will be problematic and it can't cover http servers what already exist (think about uwsgi, nginx, httpd, ...) - they usually start using configuration file and then pass some control to application.

May be `Tests` directory is a better place for this `HTTPServer`?

**Comment by seabaylea** [comment]
The aim of the `HTTPServer` API is to provide a simple out of the box HTTP server, similar to what is provided as part of the core APIs in other languages. That doesn't preclude you putting a HTTP/reverse proxy server infront.

**Comment by aleksey-mashanov** [comment]
Out of the box http server is good feature but it should be placed in a separate module to not inject additional dependencies into API's one.

Some more questions:
* Do we really need `HTTPServing` protocol in this case?
* Which one sync or async (both?) http server should come out of the box?

**Comment by seabaylea** [comment]
Yes, the `HTTPServing` protocol is somewhat superfluous given that its really a concrete implementation.

The intent is to provide an async server out of the box. Its hard to argue that the current and future concurrency model for Swift is not async - Dispatch is provided as part of the existing Swift toolchains, and large swathes of the available APIs are async based. Additionally both the Swift 5 manifesto and the recent async/await proposal points to that as the ongoing direction.

**Comment by helje5** [comment]
> Its hard to argue that the current and future concurrency model for Swift is not async

I agree with that, if you are willing to add 'default' ;-) What is wrong with using Swift to write Apache modules, Android stuff, Outlook plugins, or whatever. Nothing. There is little use in tying Swift to a specific setup.

> Dispatch is provided as part of the existing Swift toolchains

And those are super limited. Even the Linux implementation is, well, you probably know the issues much better than I do ;-)

Dispatch is in fact a major issue with secondary toolchains (BSD? Raspi? Other IoT etc).

Also specifically for Dispatch, that Tesla^d^d^d^d Google guy on [Concurrency in Swift](https://gist.github.com/lattner/31ed37682ef1576b16bca1432ea9f782):

> The one problem I anticipate with GCD is that it doesn't scale well enough: server developers in particular will want to instantiate hundreds of thousands of actors in their application

That is something which is bothering me as well (but may not affect the core choice for async, i.e. one could use libuv too).

In general my opinion is that having a basic http server lib is nice in any case (I really don't expect that serious server frameworks are going to use this for plenty of reasons). So maybe lets just get this done, but lets try to make it so, that other implementations can still use the same or similar API.

**Comment by seabaylea** [comment]
Thanks Helge. I'm going to merge this as-is: the remaining questions are around the need for the `HTTPServing` protocol and whether or not to mark as `final`, which I think we can deal with separately.

**Comment by weissi** [comment]
@seabaylea 

> Its hard to argue that the current and future concurrency model for Swift is not async

but from what I can tell the current implementation is synchronous and blocking. Will this change soon? I think we should definitely do that before pitching to swift-evo.

**Comment by seabaylea** [comment]
@weissi that's the intent, yes - any help you can provide would be greatly appreciated!

**Pull request merged**